cpp/src/io/utilities/datasource.cpp

```diff
@@ -199,6 +199,47 @@ class direct_read_source : public file_source {
   }
 };
 
+/**
+ * @brief Implementation class for reading from a device buffer source
+ */
+class device_buffer_source : public datasource {
+ public:
+  explicit device_buffer_source(cudf::device_span<uint8_t> d_buffer) : _d_buffer{d_buffer} {}
+
+  size_t host_read(size_t offset, size_t size, uint8_t* dst) final
+  {
+    CUDF_FAIL("Host read shouldn't be used with device_buffer_source");
+  }
+
+  std::unique_ptr<buffer> host_read(size_t offset, size_t size) final
+  {
+    CUDF_FAIL("Host read shouldn't be used with device_buffer_source");
+  }
+
+  [[nodiscard]] bool supports_device_read() const final { return true; }
+
+  size_t device_read(size_t offset, size_t size, uint8_t* dst, rmm::cuda_stream_view stream) final
+  {
+    CUDF_FAIL("TODO: should this be invoked?");
+    // TODO: to be finished with deep copy
```

Comment:
darn, forgot that we generally read into a pre-allocated buffer. Not really a no-op, but should still be much faster than H2D.
I guess this one should be implemented as sync D2D copy on the given stream.